src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy

```
@@ -20,153 +20,89 @@ import org.eclipse.jgit.api.DescribeCommand
+            group = 'Versioning'
+            description = 'Prints the project\'s configured version to standard out'
+            doLast {
+                println project.version
+            }
+        }
+    }
+
     @Memoized
-    private File gitDir(Project project) {
-        return getRootGitDir(project.rootDir)
+    private VersionDetails versionDetails(Project project) {
+        String description = gitDescribe(project)
+        String hash = gitHash(project)
+        String branchName = gitBranchName(project)
+        boolean isClean = isClean(project)
+
+        return new VersionDetails(description, hash, branchName, isClean)
     }
 
     @Memoized
     private Git gitRepo(Project project) {
-        return Git.wrap(new FileRepository(gitDir(project)))
+        File gitDir = GitCli.getRootGitDir(project.rootDir);
+        return Git.wrap(new FileRepository(gitDir))
     }
 
     @Memoized
-    private String gitDesc(Project project) {
+    private String gitDescribe(Project project) {
         // verify that "git" command exists (throws exception if it does not)
-        verifyGitCommandExists()
+        GitCli.verifyGitCommandExists()
 
         Git git = gitRepo(project)
         try {
             // back-compat: the JGit "describe" command throws an exception in repositories with no commits, so call it
             // first to preserve this behavior in cases where this call would fail but native "git" call does not.
             new DescribeCommand(git.getRepository()).call()
 
-            String version = runGitCommand(project.rootDir, "describe", "--tags", "--always", "--first-parent") ?: UNSPECIFIED_VERSION
-            boolean isClean = git.status().call().isClean()
-            return version + (isClean ? '' : '.dirty')
+            return GitCli.runGitCommand(project.rootDir, "describe", "--tags", "--always", "--first-parent")
         } catch (Throwable t) {
-            return UNSPECIFIED_VERSION
+            return null
```

not a big fan of the nulls everywhere. isn't reasonable to just fail instead of returning nulls (both here and below)?

---

I'm matching behaviour here - when these commands fail, we output the string 'unspecified'.